London | SDC-Nov-2025 | Ikenna Agulobi | Sprint 3| Implement shell tools

ike-agu opened the pull request:
<!--

You must title your PR like this:

Region | Cohort | FirstName LastName | Sprint | Assignment Title

For example,

London | 25-ITP-May | Carol Owen | Sprint 1 | Alarm Clock

Fill in the template below - remove any sections that don't apply.

Complete the self checklist - replace each empty box in the checklist [ ] with a [x].

Add the label "Needs Review" and you will get review.

Respond to volunteer reviews until the volunteer marks it as "Complete".

-->

## Learners, PR Template

Self checklist

- [x] I have titled my PR with Region | Cohort | FirstName LastName | Sprint | Assignment Title
- [x] My changes meet the requirements of the task
- [x] I have tested my changes
- [x] My changes follow the [style guide](https://curriculum.codeyourfuture.io/guides/reviewing/style-guide/)

## Changelist

This PR contains the completed implementations for the Sprint 3 exercises:
- cat command — file output with optional line numbering
- ls command — directory listing with -1 and -a support
- wc command — line, word, and character counting

ike-agu committed:
Intalled commader library; Create cat.js file

ike-agu committed:
implement cat command with -n and -b flag support

ike-agu committed:
Implement ls command to list files in directory with -1 flag and -a flag

ike-agu committed:
Implement wc command to count  lines, words and characters

eliza-an commented on implement-shell-tools/cat/cat.js:
```diff
@@ -0,0 +1,44 @@
+let lineNumber = 1;
+
+// Loop over every filepath in args
+args.forEach(async (filepath) => {
```
Although this works it's not the most readable of implementations. Would your peer be able to read this easily and understand what is happening? Likely no! So it's super important our code is as readable as poossible, not necissarily as short as possible!
eg: 

for (const path of filepaths) {
    const filesContent = await fs.readFile(path, "utf-8");
    this will make it easier for everyone to understand

ike-agu replied:
Thanks for pointing this out and recommending  the for...of loop instead of the forEach loop. The code is much cleaner and easier to read after I implemented the fix.

eliza-an commented on implement-shell-tools/cat/cat.js:
```diff
@@ -0,0 +1,44 @@
+  const content = await fs.readFile(filepath, "utf8");
+  const lines = content.split("\n");
+
+  lines.forEach((line) => {
```
similarly here the arrow function unnecisarily complicates a simple bit of code. Using the example i gave you above, how would you rewrite this?

ike-agu replied:
I used the for....of loop to refactor my code and it's much cleaner and readable. Thank you🙂

eliza-an commented on implement-shell-tools/wc/wc.js:
```diff
@@ -0,0 +1,80 @@
+import process from "node:process";
+
+program
+  .name("wc command")
```
this name is in a different format to your other names, not the end of the world, but consistency is always encouraged!

eliza-an commented on implement-shell-tools/wc/wc.js:
```diff
@@ -0,0 +1,80 @@
+  .option("-l,", "show number of lines only")
+  .option("-w,", "show number of words only")
+  .option("-c,", "show number of characters only")
+  .argument("[path...]", "The file path to process");
```
It is important to note that [] brackets mean the parameter is options and the <> brackets mean it is required. Remember this for my next comment below

eliza-an commented on implement-shell-tools/wc/wc.js:
```diff
@@ -0,0 +1,80 @@
+program
+  .name("wc command")
+  .description("Implement wc command to count words and lines")
+  .option("-l,", "show number of lines only")
```
these commas in the brackets shouldnot exist! your code won't work then!

ike-agu replied:
Thanks for pointing out that bug. I've updated my code  with the correct flags.

eliza-an reviewed (commented):
I left a few notes for you. This is a really god first try. There are some bits you need to change. Comments are great, but your code should be so readable that comments are unnecessary ( or at least as readable as can get!). If you need comments to understand your code, you need to simplify it so that you can read it easier! Well done overall. A good try, but some things need to be changed so your code works appropriately!

ike-agu committed:
refactor: improve readability by Deleting comments and replacing forEach with for...of loops

ike-agu committed:
refactor: improve naming and simplify ls filtering logic

ike-agu committed:
fix wc flags and align behavior with Unix wc

ike-agu commented on implement-shell-tools/cat/cat.js:
```diff
@@ -0,0 +1,44 @@
+
+program.parse(process.argv);
+
+const args = program.args; //Array all file paths
```
You're absolutely right. I've updated my code. Thanks for the feedback🙂

ike-agu commented on implement-shell-tools/ls/ls.js:
```diff
@@ -0,0 +1,42 @@
+for (const directoryPath of paths) {
+  let entries;
+  try {
+    entries = await fs.readdir(directoryPath);
```
I agree, entries doesn’t say much about the code. I've updated the name to fileNames. Thanks!🙏🏽

ike-agu commented on implement-shell-tools/ls/ls.js:
```diff
@@ -0,0 +1,42 @@
+  }
+
+  if (!opts.allFiles) {
+    entries = entries.filter((name) => !name.startsWith("."));
```
I’ve implemented the fix in a single line using filter.  The condition is now clearer and more readable. Thanks!

ike-agu commented on implement-shell-tools/wc/wc.js:
```diff
@@ -0,0 +1,80 @@
+
+if (paths.length === 0) {
+  console.error("wc: no file specified");
+  process.exit(1);
```
Thank you for highlighting the error. I've implemented the fix so that if a path is not specified, it defaults to the current path.

ike-agu commented on implement-shell-tools/wc/wc.js:
```diff
@@ -0,0 +1,80 @@
+    continue; // move on to next path
+  }
+
+  const lineCount = content.split("n").length;
```
I forgot the back slash  '\' which  means my code splits on letter n instead of splitting on a new line like so '\n'.  I've implemented the fix. Thanks🙏

ike-agu commented on implement-shell-tools/wc/wc.js:
```diff
@@ -0,0 +1,80 @@
+    continue;
+  }
+
+  if (opts.l) {
```
I've implement a fix so that the formatting aligns with wc command. The code now works as expected. Thanks 🙏

ike-agu commented:
Hi @eliza-an , apologies for circling back to this PR after such a long time. I should have addressed the feedback earlier.
Thank you again for taking the time to review it and for the valuable feedback. 
I’ve now implemented all the requested changes.
When you have a moment, could you please take another look? Thank you!

illicitonion commented on implement-shell-tools/cat/cat.js:
```diff
@@ -0,0 +1,42 @@
+  const lines = fileContent.split("\n");
+
+  for (const line of lines) {
+    if (options.numberAllLines) {
```
This works, but something to think about for the future - here you're using if conditions to switch behaviours completely. But _most_ of the behaviours you want in all of the cases are actually the same (or at least, the same shape). So often we'll structure this code more like:

```js
let prefix = "";
if (options.numberAllLines || (options.numberNonEmptyLines && line.trim() !=== "") {
  prefix = `${lineNumber} `;
  lineNumber++;
}
console.log(`${prefix} ${line}`
```

This makes it more clear what's the _same_ in all of the cases, and what's _different_ between them. It also means that if we add more small changes (e.g. an option to put something at the end of the line) these compose well, rather than exploding into lots of separate options / ifs.

illicitonion commented on implement-shell-tools/ls/ls.js:
```diff
@@ -0,0 +1,39 @@
+  }
+
+  fileNames = fileNames.filter(
+    (file) => options.allFiles || !file.startsWith("."),
```
This is a really clear way of expressing the constraint! 🎉

illicitonion commented on implement-shell-tools/wc/wc.js:
```diff
@@ -0,0 +1,63 @@
+
+  const output = [];
+
+  if (!opts.lines && !opts.words && !opts.chars) {
+    output.push(lineCount, wordCount, charCount);
+  } else {
+    if (opts.lines) output.push(lineCount);
+    if (opts.words) output.push(wordCount);
+    if (opts.chars) output.push(charCount);
+  }
+  console.log(`${output.join(" ")} ${filePath}`);
+}
```
This code is very similar to the code for accumulating totals. Extracting a function for outputting a line may be useful here (or may not - try it - it's not clear for this level of similarity/difference which would be more clear :))
